lib/Dialect/BaseModelica/Transforms/CallCSE.cpp

```diff
@@ -17,80 +19,103 @@ class CallCSEPass final : public impl::CallCSEPassBase<CallCSEPass> {
 
 private:
   mlir::LogicalResult processModelOp(ModelOp modelOp);
-
-  /// Replace all calls in the equivalence group with gets to a generated
-  /// variable. The variable will be driven by an equation derived from the
-  /// first call in the group.
-  ///
-  /// One variable and driver equation will be emitted per result,
-  /// if the call is to a function with multiple result values.
-  void emitCse(llvm::SmallVectorImpl<CallOp> &equivalenceGroup, ModelOp modelOp,
-               DynamicOp dynamicOp, mlir::SymbolTable &symbolTable,
-               mlir::RewriterBase &rewriter);
 };
 
-/// Get all call operations in the model.
-void collectCallOps(ModelOp modelOp, llvm::SmallVectorImpl<CallOp> &callOps) {
+bool haveSameInductionRanges(
+    llvm::SmallVectorImpl<EquationInstanceOp> &instanceOps) {
+  std::optional<IndexSet> seenSpace;
+  for (EquationInstanceOp instanceOp : instanceOps) {
+    IndexSet iterationSpace = instanceOp.getIterationSpace();
+    if (!seenSpace.has_value()) {
+      seenSpace = iterationSpace;
+    } else if (seenSpace.value() != iterationSpace) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/// Reduces instances to a single one per template.
+/// Completely filters out instances where induction variable ranges do not
+/// match.
+llvm::SmallVector<EquationInstanceOp>
+filterInstanceOps(llvm::SmallVectorImpl<EquationInstanceOp> &instanceOps) {
+  llvm::SmallVector<EquationInstanceOp> filteredInstanceOps;
+  llvm::DenseMap<EquationTemplateOp, llvm::SmallVector<EquationInstanceOp>>
+      instancesByTemplate;
+  for (EquationInstanceOp equationOp : instanceOps) {
```

Comment:
You can ask for the users of the `equation_template` result value.

Reply:
I'm not sure I understand how that would help here, as I already have all the instances, and just want to group them by template

Reply:
True